unified_addrs.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+import sys; assert sys.version_info[0] >= 3, "Python 3 required."
+
+import math
+import struct
+
+from pyblake2 import blake2b
+from bech32m import bech32_encode, bech32_decode, convertbits, Encoding
+
+from tv_output import render_args, render_tv, Some
+from tv_rand import Rand
+from f4jumble import f4jumble, f4jumble_inv
+import sapling_key_components
+import orchard_key_components
+
+def encode_unified(receivers):
+    orchard_receiver = b""
+    if receivers['orchard']:
+        orchard_receiver = b"".join([b"\x03", len(receivers['orchard']).to_bytes(1, byteorder='little'), receivers['orchard']])
+
+    sapling_receiver = b""
+    if receivers['sapling']:
+        sapling_receiver = b"".join([b"\x02", len(receivers['sapling']).to_bytes(1, byteorder='little'), receivers['sapling']])
+
+    t_receiver = b""
+    if receivers['transparent']:
+        t_receiver = b"".join([b"\x01", len(receivers['transparent']).to_bytes(1, byteorder='little'), receivers['transparent']])
+
+    r_bytes = b"".join([orchard_receiver, sapling_receiver, t_receiver, bytes(16)])
+    converted = convertbits(f4jumble(r_bytes), 8, 5)
+    return bech32_encode("u", converted, Encoding.BECH32M)
+
+def decode_unified(addr_str):
+    (hrp, data, encoding) = bech32_decode(addr_str)
+    assert hrp == "u" and encoding == Encoding.BECH32M
+
+    decoded = f4jumble_inv(bytes(convertbits(data, 5, 8, False)))
+    suffix = decoded[-16:]
+    # check trailing zero bytes
+    assert suffix == bytes(16)
+    decoded = decoded[:-16]
+
+    s = 0
+    acc = []
+    result = {}
+    for b in decoded:
+        if s == 0:
+            receiver_type = b
+            assert [1, 2, 3].count(b) > 0, "receiver type " + str(b) + " not recognized"
+            s = 1
+        elif s == 1:
+          receiver_len = b
+          expected_len == {1: 20, 2: 43, 3: 43}.get(receiver_type)
+            if expected_len is not None:
+                assert receiver_len == expected_len, "incorrect receiver length"
+        elif s == 2:
```

Comment:
How is this case ever triggered? `s` is only ever set to either `0` or `1`.

EDIT: Ah, the `s = 2` state transition was accidentally removed in the second commit. Add it back in:
```suggestion
            s = 2
        elif s == 2:
```